Add more ActivitySource docs (#24324)
Azure/azure-sdk-for-net: sdk/core/Azure.Core/samples/Diagnostics.md

When targeting .NET Standard 2.1, .NET Core 2.2, or newer, you might instead use `e.TimeStamp` to log the time the event was written instead of rendered, like above. It's in UTC format, so if you want to log the local time like in the example call `ToLocaleTime()` first.
For help diagnosing multi-threading issues, you might also log `e.OSThreadId` which is also available on those same targets.

## ActivitySource support

Azure SDKs released after October 2021 have experimental ActivitySource support. ActivitySource is a simplified way to create and listen to activities added in .NET 5.
Azure SDKs produce the following kinds of Activities:

- *HTTP calls*: every HTTP call originating from Azure SDKs
- *client method calls*: for example, `BlobClient.DownloadTo` or `SecretClient.StartDeleteSecret`.

> - *messaging events*: EventHubs and ServiceBus message creation is traced and correlated with its sending, receiving, and processing.

Reviewer: ```suggestion
- *messaging events*: Event Hubs and Service Bus message creation is traced and correlated with its sending, receiving, and processing.
```

To be pedantic, Event Hubs messages are not correlated for receiving.  It was defined as optional and "low-value" in the spec, and @lmolkova has offered some strong rationale why adding it isn't a good idea.

I'm not sure that it is important to make that distinction here, but I wanted to mention it, just in case.

Author: Hm, we still create spans for Receive client calls, don't we?

Reviewer: In Service Bus yes,  in Event Hubs no.

Author: Huh, TIL :)

> ActivitySource support is experimental.
> We reserve the right to change the shape of Activities in the future, including things like:
> - kinds of operations we track
> - relationships between telemetry spans
> - attributes attached to telemetry spans

Reviewer: This could be read a bit harshly.  We may want to consider softening to something like:

```suggestion
Because `ActivitySource` support is experimental, the shape of Activities may change in the future without notice.  This includes:

- the kinds of operations that are tracked
- relationships between telemetry spans
- attributes attached to telemetry spans
```

> Do one of the following steps to enable ActivitySource support:

Reviewer 2: ```suggestion
ActivitySource support can be enabled through either of these two steps:
```

> - Set `AZURE_EXPERIMENTAL_ENABLE_ACTIVITY_SOURCE` to `true`.

Reviewer 2: ```suggestion
- Set the `AZURE_EXPERIMENTAL_ENABLE_ACTIVITY_SOURCE` environment variable to `true`.
```

> - Set `Azure.Experimental.EnableActivitySource` context switch to true in your application code:

Reviewer 2: ```suggestion
- Set the `Azure.Experimental.EnableActivitySource` context switch to true in your application code:
```

If your application already uses ApplicationInsights, automatic collection of Azure SDK traces is supported since version `2.12.0` ([Microsoft.ApplicationInsights on NuGet](https://www.nuget.org/packages/Microsoft.ApplicationInsights/)).

To setup ApplicationInsights tracking for your application follow the [Start Monitoring Application](https://docs.microsoft.com/azure/azure-monitor/learn/dotnetcore-quick-start) guide.

### OpenTelemetry with Azure Monitor, Zipkin and others

OpenTelemetry relies on ActivitySource to collect distributed traces. Follow steps in [ActivitySource support](#ActivitySource support) section before proceeding to OpenTelemetry configuration.

Follow the [OpenTelemetry configuration guide](https://github.com/open-telemetry/opentelemetry-dotnet#configuration-with-microsoftextensionsdependencyinjection) to configure collecting distribute tracing event collection using the OpenTelemetry library.

> 

Reviewer 3: Can we add something for potential tracing vendors or advanced users looking for the details on what we emit? Something along the lines of 

"If you want to process, filter, or query Azure SDK telemetry spans or build Azure-specific visualizations, here are the [semantic conventions](https://github.com/Azure/azure-sdk/blob/main/docs/tracing/distributed-tracing-conventions.yml) Azure SDK follows."

Author: Can do. The only thing I worry about is the mixed messaging. Do we treat https://github.com/Azure/azure-sdk/blob/main/docs/tracing/distributed-tracing-conventions.yml as experimental?

Reviewer 3: We're not going to change them ever, but we're going to switch to otel conventions and stop supporting ours. 
They are also our contract with AzMon until OTel stabilizes, they are versioned, follow semver, and didn't change in the past year or two.

So I'd say they are not experimental, but we'll have a new major version in the next year switching to something standard hopefully everyone supports by then.

Reviewer 3: it makes me think that we should include schema_url of that spec on the spans, so processing could be versioned

Author: >So I'd say they are not experimental, but we'll have a new major version in the next year switching to something standard

Yeah, that sounds like a breaking change and I don't really want customers to start depending on something we expect to remove in a year.

Reviewer 3: Ok, so do you think we shouldn't mention them at all or say they're to play with but not depend upon?

Application Insights, a feature of Azure Monitor, is an extensible Application Performance Management (APM) service for developers and DevOps professionals. Use it to monitor your live applications. It will automatically detect performance anomalies, and includes powerful analytics tools to help you diagnose issues and to understand what users actually do with your app

> If your application already uses ApplicationInsights, automatic collection of Azure SDK traces is supported since version `2.12.0`.

Reviewer: > If your application already uses ApplicationInsights, automatic collection of Azure SDK traces is supported since version `2.12.0`.

Should we specify the package name or link to NuGet?
